### PR TITLE
[Access] Add circuit breaker for unavailable Collection/Execution nodes

### PR DESCRIPTION
**Pull request by Guitarheroua**
https://github.com/onflow/flow-go/issues/3129

### Context
This pull request adds [Circuit Breaker pattern](https://learn.microsoft.com/en-us/previous-versions/msp-n-p/dn589784(v=pandp.10)) to handle the availability and reachability of Collection and Execution nodes. I use the existing package [gobreaker](https://github.com/sony/gobreaker) from Sony, which provides all the necessary functionality and necessary configurations.

Every call now passes through the [Circuit Breaker's Execute method](https://github.com/sony/gobreaker/blob/f4dfee680910ab45229b9737850f2f5094f13366/gobreaker.go#L228)(encapsulate [the invoker](https://github.com/onflow/flow-go/blob/6aee5ddaabc5ee6bb7edab5e9c0987b270a0bea8/engine/access/rpc/backend/connection_factory.go#L270) within a UnaryInterceptor) to track calls success/failure. This implies that each node connection has its own individual Circuit Breaker and should never invalidate a connection if CB is enabled. The node selection is also improved as instead of using 3 samples when CB is enabled, it uses all available connections, that would be handled by CB.

### CODE REVIEW

File: engine/access/rpc/engine.go [hunk 1]
```diff
@@ -171,6 +172,12 @@ func NewBuilder(log zerolog.Logger,
 		MaxMsgSize:                config.MaxMsgSize,
 		AccessMetrics:             accessMetrics,
 		Log:                       log,
+		CircuitBreakerConfig:      config.CircuitBreakerConfig,
+	}
+
+	circuitBreakerEnabled := false
+	if config.CircuitBreakerConfig != nil {
```
Comment (peterargue) [inline engine/access/rpc/engine.go:179]: rather than a fallback, I'm in favor of requiring that a `config.CircuitBreakerConfig` is provided, and adding a default value into the node builders

File: engine/access/rpc/backend/connection_factory.go [hunk 2]
```diff
@@ -84,6 +111,20 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
...
+
+	cbInterceptor := cf.createCircuitBreakerInterceptor()
+	if cbInterceptor != nil {
+		connInterceptors = append(connInterceptors, cbInterceptor)
+	}
+
+	ciInterceptor := cf.createClientInvalidationInterceptor(address, clientType)
+	if ciInterceptor != nil {
+		connInterceptors = append(connInterceptors, ciInterceptor)
+	}
```
Comment (peterargue) [inline engine/access/rpc/backend/connection_factory.go:124]: how about moving the checks here instead of returning nil? it make the code clearer
```suggestion
	if cf.CircuitBreakerConfig.Enabled {
		connInterceptors = append(connInterceptors, cf.createCircuitBreakerInterceptor())
	} else {
		connInterceptors = append(connInterceptors, cf.createClientInvalidationInterceptor(address, clientType))
	}
```

**Review by peterargue — COMMENTED** [review]
Looks pretty good. I added a couple cosmetic comments. I haven't had a chance to look through all of the tests yet, but figured I'd provide some feedback sooner.

The only concern I have with this approach is that we may send requests to all of the nodes in extreme cases. For collection nodes, that could be 30-40 nodes.

I think there still needs to be some way to limit the number of actual calls made. One half-baked idea is to create a retires type package where you can pass it the list of execution nodes and a callback that takes a node and returns an error. It will internally create the selector, then call the callback for each node checking if the returned error was from the circuit breaker or the node. It then makes up to `n` calls to actual nodes, and return a multierror.

There may be better options, but we should enforce some limit.

File: engine/access/rpc/backend/connection_factory.go [hunk 3]
```diff
@@ -253,8 +294,106 @@ func getGRPCAddress(address string, grpcPort uint) (string, error) {
...
+func (cf *ConnectionFactoryImpl) createCircuitBreakerInterceptor() grpc.UnaryClientInterceptor {
+	if cf.CircuitBreakerConfig != nil && cf.CircuitBreakerConfig.Enabled {
+		circuitBreaker := gobreaker.NewCircuitBreaker(gobreaker.Settings{
+			// The restore timeout is defined to automatically return the circuit breaker to the HalfClose state.
+			Timeout: cf.CircuitBreakerConfig.RestoreTimeout,
+			ReadyToTrip: func(counts gobreaker.Counts) bool {
+				// The number of maximum failures is checked before the circuit breaker goes to the Open state.
+				return counts.ConsecutiveFailures >= cf.CircuitBreakerConfig.MaxFailures
+			},
+			MaxRequests: cf.CircuitBreakerConfig.MaxRetries,
```
Comment (peterargue) [inline engine/access/rpc/backend/connection_factory.go:362]: This naming is a little confusing. `MaxRequests` implies that it will allow up to (1 by default) requests. You've renamed it to `MaxRetries` which implies it should allow 2 requests, the initial and a single retry. Reading the doc, it seems like the library chose a confusing name. If that's the case, can you leave a comment here for future readers

**Comment by peterargue** [comment]
> Could you please explain to me in more detail what could possibly go wrong? Maybe I don't see the entire picture here.

By default, there's a 3 second timeout per node. Imagine the case where the network is under load, and collection nodes are slow to respond. Each time a transaction is submitted, it could in the worst case be sent to 40 nodes. This ties resources up on the AN and increases the load on collection nodes. Remember there are >100 Access nodes.

> This limitation already exists in Sony gobreaker; however, I haven't provided the option to configure it via CLI flags, so the solution would be to add it.
> 
> It's called [MaxRequests](https://github.com/sony/gobreaker/blob/f4dfee680910ab45229b9737850f2f5094f13366/gobreaker.go#L83). You can find the documentation for it here: `MaxRequests defines the maximum number of requests allowed to pass through when the CircuitBreaker is in the half-open state. If MaxRequests is set to 0, the CircuitBreaker allows only 1 request`.
> 
> The idea behind this configuration is that when the circuit breaker tries to restore itself after a timeout, it will make `MaxRequests` number of requests to determine whether it should transition back to the `Open` state and block requests for the timeout period, or return to the `Closed` state and allow requests again
> 
> If this, is what you mean - I could easily implement it.

The docs for `MaxRequests` are a little confusing. Looking at the code, it seems like it serves 2 purposes:
1. Limit the number of concurrent requests to a particular backend while the CB is in the half open state
2. Specify how many successful requests are needed before marking the CB closed.

That's a little different from what I'm talking about. I'm saying that with the current implementation, given a set of backends we will send requests to each one until either we get a successful response or run out of nodes. I think we need is a way to limit the number of actual requests sent to backends.

for example, say we had 30 nodes. The first 5 we attempted all had open CB, so no requests were sent. The next 3 all failed for various reasons. After sending those 3 requests, we should stop and return an error to the client.

File: integration/tests/access/access_circuit_breaker_test.go [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+	// 2. Get the Access Node container and client
+	accessContainer := s.net.ContainerByName(testnet.PrimaryAN)
+
+	// Check if access node was created with circuit breaker flags
+	require.True(s.T(), accessContainer.IsFlagSet("circuit-breaker-enabled"))
+	require.True(s.T(), accessContainer.IsFlagSet("circuit-breaker-restore-timeout"))
+	require.True(s.T(), accessContainer.IsFlagSet("circuit-breaker-max-failures"))
+
+	accessClient, err := accessContainer.TestnetClient()
+	assert.NoError(s.T(), err, "failed to get collection node client")
```
Comment (peterargue) [inline integration/tests/access/access_circuit_breaker_test.go:123]:
```suggestion
	assert.NoError(s.T(), err, "failed to get access node client")
```

File: integration/tests/access/access_circuit_breaker_test.go [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+		duration := time.Since(start)
+		defer cancel()
+
+		return duration, err
+	}
+
+	// Try to send the transaction for the first time. It should wait at least the timeout time and return Unknown error
+	duration, err := sendTransaction(ctx, signedTx)
+	assert.Equal(s.T(), codes.Unknown, status.Code(err))
+	assert.GreaterOrEqual(s.T(), requestTimeout, duration)
```
Comment (peterargue) [inline integration/tests/access/access_circuit_breaker_test.go:179]: doesn't this mean that `duration <= requestTimeout`? From the comment, it sounds like it should be the greater

File: integration/tests/access/access_circuit_breaker_test.go [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+		err := accessClient.SendTransaction(childCtx, tx)
+		duration := time.Since(start)
+		defer cancel()
+
+		return duration, err
+	}
+
+	// Try to send the transaction for the first time. It should wait at least the timeout time and return Unknown error
+	duration, err := sendTransaction(ctx, signedTx)
+	assert.Equal(s.T(), codes.Unknown, status.Code(err))
```
Comment (peterargue) [inline integration/tests/access/access_circuit_breaker_test.go:178]: does the node currently return `Unknown`? It should really be `Unavailable`. If this isn't new from your code, we can update in another PR.

File: engine/access/rpc/backend/backend_test.go [hunk 7]
```diff
@@ -169,7 +171,7 @@ func (suite *Suite) TestGetLatestProtocolStateSnapshot_NoTransitionSpan() {
 		epochBuilder := unittest.NewEpochBuilder(suite.T(), state)
 		// build epoch 1
 		// blocks in current state
-		// P <- A(S_P-1) <- B(S_P) <- C(S_A) <- D(S_B) |setup| <- E(S_C) <- F(S_D) |commit|
+		// P <- A(S_P-1) <- B(S_P) <- C(S_A) <- D(S_B) |setup| <- E(S_C) <- NodeAction(S_D) |commit|
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_test.go:174]: were these changes intended?

File: cmd/access/node_builder/access_node_builder.go [hunk 8]
```diff
@@ -744,6 +753,14 @@ func (builder *FlowAccessNodeBuilder) extraFlags() {
...
 			}
 		}
+		if builder.rpcConf.CircuitBreakerConfig.Enabled {
+			if builder.rpcConf.CircuitBreakerConfig.MaxFailures == 0 {
+				return errors.New("circuit-breaker-max-failures must be greater than 0")
+			}
+			if builder.rpcConf.CircuitBreakerConfig.MaxRequests == 0 {
+				return errors.New("circuit-breaker-max-requests must be greater than 0")
+			}
+		}
```
Comment (peterargue) [inline cmd/access/node_builder/access_node_builder.go:763]: can you also check that `RestoreTimeout` is >0

**Review by peterargue — DISMISSED** [review]

**Review by peterargue — COMMENTED** [review]
left a few small comments, but this is pretty much ready to go

**Comment by peterargue** [comment]
looks like there's a panic in the circuit breaker test

File: engine/access/rpc/backend/node_selector.go [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+	return &MainNodeSelector{
+		nodes: nodes,
+		index: 0,
+	}, nil
+}
+
+var _ NodeSelector = (*MainNodeSelector)(nil)
+
+// MainNodeSelector is a specific implementation of the node selector.
+type MainNodeSelector struct {
```
Comment (durkmurder) [inline engine/access/rpc/backend/node_selector.go:47]: Can you add a constructor to make sure index is properly initialized?

File: engine/access/rpc/backend/node_selector.go [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+	nodes flow.IdentityList
+	index int
+}
```
Comment (durkmurder) [inline engine/access/rpc/backend/node_selector.go:50]:
```suggestion
}

var _ NodeSelector = (*MainNodeSelector)(nil)

```

File: engine/access/rpc/backend/node_selector.go [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+// NodeSelector is an interface that represents the ability to select node identities that the access node is trying to reach.
+// It encapsulates the internal logic of node selection and provides a way to change implementations for different types
+// of nodes. Implementations of this interface should define the Next method, which returns the next node identity to be
+// selected. HasNext checks if there is next node available.
+type NodeSelector interface {
+	Next() *flow.Identity
+	HasNext() bool
+}
+
+// NodeSelectorFactory is a factory for creating node selectors based on factory configuration and node type.
```
Comment (durkmurder) [inline engine/access/rpc/backend/node_selector.go:21]:
```suggestion
// NodeSelectorFactory is a factory for creating node selectors based on factory configuration and node type.
// Supported configurations:
// circuitBreakerEnabled = true - nodes will be pseudo-randomly sampled and picked in-order.
// circuitBreakerEnabled = false - nodes will be picked from proposed list in-order without any changes.
```

File: engine/access/rpc/backend/connection_factory_test.go [hunk 12]
```diff
@@ -407,6 +409,172 @@ func TestConnectionPoolStale(t *testing.T) {
...
+	connectionFactory.CircuitBreakerConfig = CircuitBreakerConfig{
+		Enabled:        true,
+		MaxFailures:    1,
+		MaxRequests:    1,
+		RestoreTimeout: circuitBreakerRestoreTimeout,
+	}
+
+	// Set the connection pool cache size.
+	cacheSize := 1
+	cache, _ := lru.NewWithEvict(cacheSize, func(_, evictedValue interface{}) {
```
Comment (durkmurder) [inline engine/access/rpc/backend/connection_factory_test.go:446]: We can just new `lru.New` if we don't plan to test eviction.

**Review by durkmurder — APPROVED** [review]
Looks good, added a few stylistic comments

**Comment by peterargue** [comment]
looks like circuit breaker integration tests are failing

**Comment by peterargue** [comment]
I just merge https://github.com/onflow/flow-go/pull/4534

File: engine/access/rpc/backend/backend_scripts.go [hunk 13]
```diff
@@ -119,8 +104,8 @@ func (b *backendScripts) executeScriptOnExecutor(
 	// *DO NOT* use this hash for any protocol-related or cryptographic functions.
 	insecureScriptHash := md5.Sum(script) //nolint:gosec
 
-	// try execution on Archive nodes first
-	if len(b.archiveAddressList) > 0 {
+	// try execution on Archive nodes if there is no execution nodes found
+	if len(executors) == 0 && len(b.archiveAddressList) > 0 {
```
Comment (peterargue) [inline engine/access/rpc/backend/backend_scripts.go:108]: I think we should keep the original. the intention of this code is to override the normal behavior and execute scripts on archive nodes instead. If there are not archive node configured or if the script fails, it should fall back to use execution nodes.

File: engine/access/rpc/connection/connection_test.go [hunk 14]
```diff
@@ -50,6 +52,12 @@ func TestProxyAccessAPI(t *testing.T) {
 		unittest.Logger(),
 		connectionFactory.AccessMetrics,
 		0,
+		CircuitBreakerConfig{
+			Enabled:        false,
+			MaxFailures:    0,
+			MaxRequests:    0,
+			RestoreTimeout: 0,
+		},
```
Comment (peterargue) [inline engine/access/rpc/connection/connection_test.go:60]: nit, add a method that returns this default disabled config and call that instead of repeating this struct.

Alternatively, since 0 and false are default initial values, so you could just use
```golang
CircuitBreakerConfig{}
```

File: integration/testnet/client.go [hunk 15]
```diff
@@ -79,18 +79,18 @@ func NewClient(addr string, chain flow.Chain) (*Client, error) {
...
-	//}
-	//public := key.PublicKey(1000)
-	//publicJson, err := public.MarshalJSON()
-	//if err != nil {
-	//	return nil, fmt.Errorf("cannot marshal key json: %w", err)
-	//}
-
-	//fmt.Printf("New client with private key: \n%s\n", json)
-	//fmt.Printf("and public key: \n%s\n", publicJson)
+	json, err := key.MarshalJSON()
```
Comment (peterargue) [inline integration/testnet/client.go:82]: can you comment this back out? best not to print the keys the stdout

**Review by peterargue — APPROVED** [review]
Added a couple small comments, but looks good once those are addressed. Great work!

**Review by peterargue — APPROVED** [review]
